# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/cookie/CookieService.cpp
# Commit: 123069093c2f
# Full Hash: 123069093c2fc18e87e9ee19c2d22815bc6b2804
# Author: Valentin Gosu <valentin.gosu@gmail.com>
# Date: 2025-03-28 21:29:38
# Regressor Bug: 1951750
# File Overlap Count: 4
# Description:
#   Bug 1951750 - Handle domain argument in CookieStoreParent r=baku,edgul,cookie-reviewers
#   
#   Depends on D241966
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D240695
# ==============================================================================

diff -r f3959de161ba -r 123069093c2f netwerk/cookie/CookieService.cpp
--- a/netwerk/cookie/CookieService.cpp	Thu Mar 27 19:06:33 2025 +0000
+++ b/netwerk/cookie/CookieService.cpp	Thu Mar 27 19:06:33 2025 +0000
@@ -719,18 +719,21 @@
     return NS_ERROR_INVALID_ARG;
   }
 
-  return AddNative(aHost, aPath, aName, aValue, aIsSecure, aIsHttpOnly,
+  return AddNative(nullptr, aHost, aPath, aName, aValue, aIsSecure, aIsHttpOnly,
                    aIsSession, aExpiry, &attrs, aSameSite, aSchemeMap,
-                   aIsPartitioned, nullptr);
+                   aIsPartitioned, nullptr,
+                   [](CookieStruct&) -> bool { return true; });
 }
 
 NS_IMETHODIMP_(nsresult)
-CookieService::AddNative(const nsACString& aHost, const nsACString& aPath,
-                         const nsACString& aName, const nsACString& aValue,
-                         bool aIsSecure, bool aIsHttpOnly, bool aIsSession,
-                         int64_t aExpiry, OriginAttributes* aOriginAttributes,
-                         int32_t aSameSite, nsICookie::schemeType aSchemeMap,
-                         bool aIsPartitioned, const nsID* aOperationID) {
+CookieService::AddNative(nsIURI* aCookieURI, const nsACString& aHost,
+                         const nsACString& aPath, const nsACString& aName,
+                         const nsACString& aValue, bool aIsSecure,
+                         bool aIsHttpOnly, bool aIsSession, int64_t aExpiry,
+                         OriginAttributes* aOriginAttributes, int32_t aSameSite,
+                         nsICookie::schemeType aSchemeMap, bool aIsPartitioned,
+                         const nsID* aOperationID,
+                         const std::function<bool(CookieStruct&)>& aCheck) {
   if (NS_WARN_IF(!aOriginAttributes)) {
     return NS_ERROR_FAILURE;
   }
@@ -751,20 +754,22 @@
   NS_ENSURE_SUCCESS(rv, rv);
 
   int64_t currentTimeInUsec = PR_Now();
-  CookieKey key = CookieKey(baseDomain, *aOriginAttributes);
-
-  CookieStruct cookieData(nsCString(aName), nsCString(aValue), nsCString(aHost),
+  CookieStruct cookieData(nsCString(aName), nsCString(aValue), host,
                           nsCString(aPath), aExpiry, currentTimeInUsec,
                           Cookie::GenerateUniqueCreationTime(currentTimeInUsec),
                           aIsHttpOnly, aIsSession, aIsSecure, aIsPartitioned,
                           aSameSite, aSameSite, aSchemeMap);
 
-  RefPtr<Cookie> cookie = Cookie::Create(cookieData, key.mOriginAttributes);
+  if (!aCheck(cookieData)) {
+    return NS_ERROR_FAILURE;
+  }
+
+  RefPtr<Cookie> cookie = Cookie::Create(cookieData, *aOriginAttributes);
   MOZ_ASSERT(cookie);
 
   CookieStorage* storage = PickStorage(*aOriginAttributes);
   storage->AddCookie(nullptr, baseDomain, *aOriginAttributes, cookie,
-                     currentTimeInUsec, nullptr, VoidCString(), true,
+                     currentTimeInUsec, aCookieURI, VoidCString(), true,
                      !aOriginAttributes->mPartitionKey.IsEmpty(), nullptr,
                      aOperationID);
   return NS_OK;